aten/src/ATen/native/xnnpack/Factory.cpp

```diff
@@ -0,0 +1,44 @@
+#ifdef USE_XNNPACK
+
+#include <ATen/native/xnnpack/Factory.h>
+#include <ATen/native/utils/Allocator.h>
+
+namespace at {
+namespace native {
+namespace xnnpack {
+namespace internal {
+
+Tensor empty_with_tail_padding(
+    const IntArrayRef size,
+    const TensorOptions& options,
+    const c10::MemoryFormat memory_format) {
+  AT_ASSERT(options.device().is_cpu());
+
+  static GuardingAllocator<0u, XNN_EXTRA_BYTES> allocator;
+
+  const int64_t nelements = prod_intlist(size);
+  const caffe2::TypeMeta dtype = options.dtype();
+
+  c10::intrusive_ptr<c10::TensorImpl> tensor_impl =
```

Comment:
I don't really know what's going on here.  Was this reviewed by someone qualified?

Reply:
this looks correct to me except in the case where you wouldn't actually dispatch to this function.  I think you'd need asserts that the layout is strided and the dtype is normal (i.e. not quantized and not complex) and that the memory isn't pinned.

But this seems a little strange because:
1) we should probably make this easier for you, because you are just passing in a different allocator (which seems like a reasonable thing to do in C++ land)
2) going through TensorOptions seems to add more pain than it's worth (and we are in the process of changing this -- for example, memory_format will be folded into it eventually).

So, for now, I'd just recommend killing the TensorOptions and passing the dtype and memory_format only.  This is assuming that my understanding is correct and you'd never actually want to use something like this for CUDA/GPU.

Reply:
CC @yf225 in case we have a nice C++ API for this and I just missed it.

Also, calling `resize_` instead of `set_sizes_contiguous` and `empty_tensor_restride` might be more maintainable (take it or leave it).

Reply:
Will change.

Reply:
To use a different allocator, it might be slightly easier to write: https://github.com/pytorch/pytorch/blob/a64d0ffe81b250889a3e6670daa9c7d07d946e32/aten/src/ATen/native/Memory.cpp#L24-L30